library/ecdh.c

```diff
@@ -91,7 +98,79 @@ int mbedtls_ecdh_compute_shared( mbedtls_ecp_group *grp, mbedtls_mpi *z,
  */
 void mbedtls_ecdh_init( mbedtls_ecdh_context *ctx )
 {
+#if defined(MBEDTLS_ECDH_LEGACY_CONTEXT)
+    memset( ctx, 0, sizeof( mbedtls_ecdh_context ) );
+
+    mbedtls_ecp_group_init( &ctx->grp );
+    mbedtls_ecp_point_init( &ctx->Q   );
+    mbedtls_ecp_point_init( &ctx->Qp  );
+#else
     memset( ctx, 0, sizeof( mbedtls_ecdh_context ) );
+#endif
+}
+
+static int mbedtls_ecdh_setup_internal( mbedtls_ecdh_context *ctx,
+                                        mbedtls_ecp_group_id grp )
+{
+    int ret;
+#if defined(MBEDTLS_ECDH_LEGACY_CONTEXT)
+    mbedtls_ecdh_context *real_ctx = ctx;
+#else
+    mbedtls_ecdh_context_mbed *real_ctx = &ctx->ctx.mbed_ecdh;
+
+    ctx->var = MBEDTLS_ECDH_VARIANT_MBEDTLS_2_0;
+    ctx->grp = grp;
+
+    mbedtls_ecp_group_init( &real_ctx->grp );
+    mbedtls_ecp_point_init( &real_ctx->Q   );
+    mbedtls_ecp_point_init( &real_ctx->Qp  );
+#endif
+
+    ret = mbedtls_ecp_group_load( &real_ctx->grp, grp );
+    if( ret != 0 )
+    {
+        mbedtls_ecdh_free( ctx );
+        return( MBEDTLS_ERR_ECP_FEATURE_UNAVAILABLE );
+    }
+
+    return( 0 );
+}
+
+/*
+ * Setup context
+ */
+int mbedtls_ecdh_setup( mbedtls_ecdh_context *ctx, mbedtls_ecp_group_id grp )
+{
+    if( ctx == NULL )
+        return( MBEDTLS_ERR_ECP_BAD_INPUT_DATA );
+
+#if defined(MBEDTLS_ECDH_LEGACY_CONTEXT)
+    return( mbedtls_ecdh_setup_internal( ctx, grp ) );
+#else
+    switch( grp )
+    {
+        default:
+            return( mbedtls_ecdh_setup_internal( ctx, grp ) );
+    }
+#endif
+}
+
+static void mbedtls_ecdh_free_internal( mbedtls_ecdh_context *ctx )
```

Comment:
How about changing the `xxx_internal` functions as follows?
```
#if defined(MBEDTLS_ECDH_LEGACY_CONTEXT)
typedef mbedtls_ecdh_context mbedtls_ecdh_context_mbed;
#endif
// …
static int xxx_internal( mbedtls_ecdh_context_mbed* ctx )
{ … }
int xxx( mbedtls_ecdh_context *ctx )
{
#if defined(MBEDTLS_ECDH_LEGACY_CONTEXT)
    return( xxx_internal( ctx ) );
#else
    switch( ctx->var ) {
        case MBEDTLS_ECDH_VARIANT_MBEDTLS_2_0:
            return( xxx_internal( &ctx->ctx.mbed_ecdh ) );
    }
#endif
}
```

This simplifies the internal functions without making the public wrappers more complex.

Reply:
I think it is a brilliant idea!